nova/virt/vmwareapi/driver.py

```diff
@@ -693,11 +695,288 @@ def detach_interface(self, context, instance, vif):
+        :param recover_method:
+            recovery method when any exception occurs.
+            expected nova.compute.manager._rollback_live_migration.
+        :param block_migration: if true, migrate VM disk.
+        :param migrate_data: a LiveMigrateData object
+
+        """
+        dest_session = self._create_dest_session(migrate_data)
+        try:
+            disk_mappings = self._get_disk_mappings(context, instance)
+            self._set_vif_infos(instance, migrate_data, dest_session)
+            self._vmops.live_migration(instance, migrate_data, disk_mappings)
+            LOG.info("Migration operation has completed", instance=instance)
+            post_method(context, instance, dest, block_migration, migrate_data)
+        except exception.InstanceNotFound:
+            # Try to get the instance on the remote session
+            try:
+                vm_util.get_vm_ref.__wrapped__(dest_session, instance)
+                # TODO(fwiesel): get the migration task
+                LOG.info("Migration operation has recovered",
+                        instance=instance)
+                post_method(context, instance, dest, block_migration,
+                    migrate_data)
+            except exception.InstanceNotFound:
+                with excutils.save_and_reraise_exception():
+                    LOG.debug("Calling live migration recover_method "
+                            "for instance: %s", instance["name"])
+                    recover_method(context, instance, dest, migrate_data)
+        except Exception:
+            LOG.exception("Failed failed due to an expection")
+            with excutils.save_and_reraise_exception():
+                LOG.debug("Calling live migration recover_method "
+                          "for instance: %s", instance["name"])
+                recover_method(context, instance, dest, migrate_data)
+
+    def _get_disk_mappings(self, context, instance):
+        targets = {}
+        for bdm in objects.BlockDeviceMappingList.get_by_instance_uuid(
+                    context, instance.uuid):
+            if not bdm.volume_id or not bdm.connection_info:
+                continue
+
+            connection_info = jsonutils.loads(bdm.connection_info)
+            data = connection_info.get('data', {})
+            datastore = data.get("datastore") or \
+                            data.get("ds_ref_val")
+
+            if not datastore:
+                continue
+
+            target = {
+                "datastore_ref": vim_util.get_moref(datastore, "Datastore")
+            }
+
+            if "volume" in data:
+                target["volume"] = data["volume"]
+
+            if "profile_id" in data:
+                target["profile_id"] = data["profile_id"]
+
+            targets[bdm.volume_id] = target
+
+        return targets
+
+    def _set_vif_infos(self, instance, migrate_data, session):
+        if 'vifs' in migrate_data and migrate_data.vifs:
```

Comment:
I am not sure how it is done, only that it is done that way throughout live-migration. I guess coming from `NovaObject`.
I presume it is for version compatibility. The `in` check is if that attribute even exists (instead of `hasattr`), and then people seem to prefer to do an attribute access over a dict-access.